### DISCUSSION TITLE
Get UNEXPLORED status after sometime,

### DISCUSSION THREAD

**Discussion by omidziaee** [optimatika/ojAlgo#526] [General, ojAlgo and related code]
I am running almost the same problem each 5 minutes, just some inputs to the problem changes. At first it runs smoothly and after a while it times out (I set the abort time to 100 seconds) and get the UNEXPLORED status. Is there a flag or something that I need to set at the end of each round to fully restart the solver? Any help would be appriciated.
Omid

**Comment by apete** [comment]
You have to give more details on how you do things. Are you using ExpressionsBasedModel? Which solver is used?

**Comment by omidziaee** [comment]
Yes I am using ExpressionsBasedModel and the default solver.

**Comment by omidziaee** [comment]
How do you set the solver? I assume ExpressionBasedModel selects the solver based on the problem. Am I wrong?

**Comment by apete** [comment]
Yes, `ExpressionBasedModel` selects an appropriate solver based on the problem type. There's a method `addIntegration(Integration<?>)` that allows you to add 3:d party solvers.

**Comment by omidziaee** [comment]
What does UNEXPLORED mean exactly? On document I see it is New/changed problem. If there is result and it is UNEXPLORED status after time out, does it mean it is the best solution during that time?

**Comment by apete** [comment]
It would typically mean that the solver hasn't done anything, or perhaps more specifically that it hasn't reached a point where it could set some other state like FEASIBLE, INFEASIBLE or OPTIMAL.

You can turn on debug logging for the individual solver to see what goes on:

`model.options.debug(LinearSolver.class);` (or ConvexSolver, IntegerSolver or whatever you use)

If you you do incremental updates, in-between solves, to the `ExpressionBasedModel` instance; then try to create a new model instance each time.

**Reply by omidziaee** [reply]
I changed the code to recreate the model each time (no modification to the existing model). It is a bit better but again after a while it is all UNEXPLORED. Do I need to do something with the solver too or just model is enough?

**Comment by omidziaee** [comment]
@apete I created an instance of Intermediate by getModel().prepare() and then dispose it. I see dispose method in the Intermediate has some clearing for the solver also. It helps a lot and now I almost don't see an UNEXPLORED but a few time I still get UNEXPLORED. And when it get to UNEXPLORED it gets stuck. Seems it has something to do with the cleaning up. Is there another thing that I still need to do to clean the model and the solver? Could you please shed some light on this?
Thank you,

**Reply by apete** [reply]
You have to provide a way to recreate the issue, otherwise no one can help you. To start: When a problem occurs just write the model (`ExpressionsBasedModel` instance) to disk. Then see if you can solve that model in a fresh JVM.

`model.writeTo(File);`

**Reply by omidziaee** [reply]
I did. As I said if I restart the JAVA program with the same input it gives the optimal solution. The problem is it seems something is adding up which need to be cleared. Is there anyway to clear the solve also or it is just through the Intermediate instace? My current approach is clear model (through dispose method) and also clear solver (through Intermediate instance and prepare method) I am not sure if it is a correct approach or not. Could you please shed some light on this? I also attached two classes CreateModel.java and SimulateRun.java. I am not sure if it is helpful or not but in CreateModel the model is being created and in SimulateRun the minimize method gets called to run the optimization.
[CreateModel.txt](https://github.com/optimatika/ojAlgo/files/12255045/CreateModel.txt)
[SimulateRun.txt](https://github.com/optimatika/ojAlgo/files/12255046/SimulateRun.txt)

**Reply by apete** [reply]
There is nothing that needs to be cleared.

Those files are too big for me to read. Particularly since it's not possible to compile and run the code. Believe I did see that this is a MIP model. Those may be MUCH more difficult to solve than LP models. Maybe the IntegerSolver just doesn't find any feasible solutions within your time frame.

ojAlgo's MIP solver is multithreaded and, unfortunately, not deterministic. It is possible that the solver takes different paths with different invokations. You can check its progress using:
`model.options.progress(IntegerSolver.class);`
If it is "working" but just not finding any solution then try to modify the IntegerStrategy.

**Comment by omidziaee** [comment]
The binary part is not used now. I am pretty sure something is adding up. After I put the prepare through the Intermediate instance, it is much better.

**Comment by apete** [comment]
You get a new solver every time you call minimise/maximise, and you say you do not re-use the models - I don't see what it is that could "add up".

Try disabling the presolvers:

```
ExpressionsBasedModel. clearPresolvers();
ExpressionsBasedModel. addPresolver(Presolver);
ExpressionsBasedModel.removePresolver(Presolver);
```

**Reply by omidziaee** [reply]
What should be the order? Is the following order good or not?
1- create model 2- add preserver 3- model solve 4- clear presolver 5- remove presolver

**Comment by apete** [comment]
No, there is no order like that. Adding/removing presolvers and integrations are "static" changes. Change it to how you want it to be, and leave it like that (until you want it to be different). In most use cases you don't change these things at all.

Normal usage is simply:

1. Build the model
2. Call minimise or maximise
